### PR TITLE
Add code action support for regexp

### LINKED ISSUE #38568: [Task]: Add code action support for regular expressions
### Description

The following code actions don't work with regular expressions. Identify and fix if there are any other similar code actions.

1. Create function code action

`string:RegExp regex = getRegex();`

2. Change variable type code action

`string regex2 = re ``;`

3. Change return type code action 

```
function test() returns string {

    return re ``;
}
```
4. Fill record fields code action and completion item

```
type rec record {|
    string:RegExp regex;
    string str;
|};

function test() {
    rec r = {};

}
```

### Describe your task(s)

_No response_

### Related area

-> Editor

### Related issue(s) (optional)

https://github.com/ballerina-platform/ballerina-lang/issues/38095

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #39249: [Bug]: Create variable code action is not suggesting for "re ``" expression
### Description

$subject

Screenshot 2023-01-13 at 10 21 17

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Editor

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #39457: [Task]: Giving semantic api support for regexp
### Description

When we pass a Template Expression Node which is a regexp we should get the correct TypeSymbol and Symbol for that.

function symbol() {
    var reg = re `<cursor>`;
}

### Describe your task(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #38568
Fixes #39249
Fixes #39457

## Samples
https://user-images.githubusercontent.com/46857198/216900142-cc19fd50-a002-4f18-be91-977d625813e0.mov

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -413,6 +413,10 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
         if (!PackageID.ANNOTATIONS.equals(pkgNode.packageID)) {
             initPredeclaredModules(symTable.predeclaredModules, pkgNode.compUnits, pkgEnv);
         }
+        
+        if (REGEXP.equals(pkgNode.packageID)) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:417]: Did you double-check whether loading happens correctly even without this change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 2]
```diff
@@ -1143,35 +1147,40 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
-            for (int i = index; i < compUnits.size(); i++) {
-                boolean isUndefinedModule = true;
-                String compUnitName = compUnits.get(i).name;
-                if (((BPackageSymbol) entry.symbol).compUnit.value.equals(compUnitName)) {
+            definePredeclaredModuleEntry(compUnits, predeclaredModuleEntry);
+        }
+        this.env = prevEnv;
+    }
+
+    private void definePredeclaredModuleEntry(List<BLangCompilationUnit> compUnits,
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1155]: RegExp is not a predeclared module. Hence, you some other name

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java [hunk 3]
```diff
@@ -483,6 +484,10 @@ public void updateXMLSubtypeOwners() {
         this.xmlPIType.tsymbol.owner = this.langXmlModuleSymbol;
         this.xmlTextType.tsymbol.owner = this.langXmlModuleSymbol;
     }
+    
+    public void updateRegExpTypeOwners() {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java:488]: You'll have to call this at SymbolEnter as well. Please check all places where other subtype owners are being changed

**Review by pcnfernando — CHANGES_REQUESTED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java (hunk 3 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java:488]: Already done this.

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java [hunk 4]
```diff
@@ -19,26 +19,44 @@
...
 
     @Override
     public String signature() {
-        return "regexp:" + Names.REGEXP;
+        return "regexp:" + Names.STRING_REGEXP;
+    }
+
+    @Override
+    public Optional<ModuleSymbol> getModule() {
+        if (this.module == null) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java:53]: Shall we maintain a `moduleEvaluated` here? If the module is unresolvable, it will try to resolve the module per each request

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java [hunk 5]
```diff
@@ -19,26 +19,44 @@
...
+    private ModuleSymbol module;
+
     public BallerinaRegexpTypeSymbol(CompilerContext context, BRegexpType regexpType) {
         super(context, TypeDescKind.REGEXP, regexpType);
     }
 
     @Override
     public String signature() {
-        return "regexp:" + Names.REGEXP;
+        return "regexp:" + Names.STRING_REGEXP;
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java:48]: Can we do a signature assertion to cover this?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 6]
```diff
@@ -2457,7 +2457,8 @@ public BLangNode transform(TemplateExpressionNode expressionNode) {
             case RAW_TEMPLATE_EXPRESSION:
                 return createRawTemplateLiteral(expressionNode.content(), getPosition(expressionNode));
             case REGEX_TEMPLATE_EXPRESSION:
-                return createRegExpTemplateLiteral(expressionNode.content(), getPosition(expressionNode));
+                return createRegExpTemplateLiteral(expressionNode.content(), getPosition(expressionNode),
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:2460]: You can pass just the expression node and handle the content and position at the creation method

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java [hunk 7]
```diff
@@ -270,7 +276,10 @@ private static Optional<String> getDefaultValueForType(TypeSymbol bType, boolean
                 break;
             case TABLE:
                 TypeSymbol rowType = ((TableTypeSymbol) rawType).rowTypeParameter();
-                String rowValue = getDefaultValueForTypeDescKind(rowType).orElse("");
+                String rowValue = "";
+                if (rowType.typeKind() != TypeDescKind.TYPE_REFERENCE) {
+                    rowValue = getDefaultValueForTypeDescKind(rowType).orElse("");
```
Comment (pcnfernando) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java:281]: What's this change about? Can you share a sample case

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java (hunk 5 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java:48]: I'll add.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java (hunk 7 above)
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java]: Removed that unnecessary if. I here the rowValue always should be an empty.

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java (hunk 4 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/symbols/BallerinaRegexpTypeSymbol.java]: Can you check the new changes?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 8]
```diff
@@ -413,7 +413,7 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
         if (!PackageID.ANNOTATIONS.equals(pkgNode.packageID)) {
             initPredeclaredModules(symTable.predeclaredModules, pkgNode.compUnits, pkgEnv);
         }
-
+
```
Comment (malinthar) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:416]: Unwanted formatting change

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/FunctionGenerator.java [hunk 9]
```diff
@@ -34,7 +34,7 @@
  * Function generator utilities.
  */
 public class FunctionGenerator {
-
+
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/FunctionGenerator.java:37]: Unwanted formatting change

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java [hunk 10]
```diff
@@ -269,8 +274,7 @@ private static Optional<String> getDefaultValueForType(TypeSymbol bType, boolean
                 }
                 break;
             case TABLE:
-                TypeSymbol rowType = ((TableTypeSymbol) rawType).rowTypeParameter();
-                String rowValue = getDefaultValueForTypeDescKind(rowType).orElse("");
+                String rowValue = "";
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java:277]: Is there a particular reason for this change? The row value is not always a string(` ""`). It should be based on the row type.

**Review by malinthar — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java (hunk 10 above)
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java:277]: In here with the previous approach the result will be `table [{}]`. After applying a code action the result is error nus because  we have an empty mapping constructor inside the square brackets. 

But if we provide the default value as `table []`. There will be no errors right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 11]
```diff
@@ -1143,35 +1142,40 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
-            for (int i = index; i < compUnits.size(); i++) {
-                boolean isUndefinedModule = true;
-                String compUnitName = compUnits.get(i).name;
-                if (((BPackageSymbol) entry.symbol).compUnit.value.equals(compUnitName)) {
+            defineRequiredModuleEntry(compUnits, predeclaredModuleEntry);
+        }
+        this.env = prevEnv;
+    }
+
+    private void defineRequiredModuleEntry(List<BLangCompilationUnit> compUnits,
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1150]: We can revert this change since this won't be called separately now?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java [hunk 12]
```diff
@@ -462,6 +462,7 @@ public void updateBuiltinSubtypeOwners() {
         updateIntSubtypeOwners();
         updateStringSubtypeOwners();
         updateXMLSubtypeOwners();
+        updateRegExpTypeOwners();
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/SymbolTable.java:465]: Move this before string

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java (hunk 10 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/DefaultValueGenerationUtil.java:277]: I see. +1

**Review by malinthar — DISMISSED** [review]

**Review by mohanvive — APPROVED** [review]

**Review by IMS94 — APPROVED** [review]
LGTM!

**Review by pcnfernando — APPROVED** [review]
